lenovo/thinkpad/x1/6th-gen/default.nix

```diff
@@ -0,0 +1,73 @@
+# A good source of information about how to fix the issues still
+# standing with kernel 4.6.11 is the following wiki page:
+# https://wiki.archlinux.org/index.php/Lenovo_ThinkPad_X1_Carbon_(Gen_6). The
+# TrackPoint and TouchPad issues there seem to have been fixed already.
+
+{ config, pkgs, ... }:
+{
+  imports = [
+    ../.
+  ];
+  # Give TLP service more control over battery
+  boot = {
+    extraModulePackages = with config.boot.kernelPackages; [
+      acpi_call
+    ];
+    kernelModules = [
+      "acpi_call"
+    ];
+  };
+
+  # See https://linrunner.de/en/tlp/docs/tlp-faq.html#battery
+  services.tlp.extraConfig = ''
+START_CHARGE_THRESH_BAT0=75
+STOP_CHARGE_THRESH_BAT0=80
+'';
+
+  # Temporary fix for cpu throttling issues visible in the kernel log
+  # (journalctl -k) by setting the same temperature limits used by
+  # Window$
+  # See https://wiki.archlinux.org/index.php/Lenovo_ThinkPad_X1_Carbon_(Gen_6)#Power_management.2FThrottling_issues
+  systemd.services.cpu-throttling = {
+    enable = true;
+    description = "Sets the offset to 3 °C, so the new trip point is 97 °C";
+    documentation = [
+      "https://wiki.archlinux.org/index.php/Lenovo_ThinkPad_X1_Carbon_(Gen_6)#Power_management.2FThrottling_issues"
+    ];
+    path = [ pkgs.msr-tools ];
+    script = "wrmsr -a 0x1a2 0x3000000";
+    serviceConfig = {
+      Type = "oneshot";
+    };
+    wantedBy = [
+      "timers.target"
+    ];
+  };
+
+  systemd.timers.cpu-throttling = {
+    enable = true;
+    description = "Set cpu heating limit to 97 °C";
+    documentation = [
+      "https://wiki.archlinux.org/index.php/Lenovo_ThinkPad_X1_Carbon_(Gen_6)#Power_management.2FThrottling_issues"
+    ];
+    timerConfig = {
+      OnActiveSec = 60;
+      OnUnitActiveSec = 60;
+      Unit = "cpu-throttling.service";
+    };
+    wantedBy = [
+      "timers.target"
+    ];
+  };
+
+  # Enable S3 suspend state: you have to manually follow the
+  # instructions shown here: https://delta-xi.net/#056 in order to
+  # produce the ACPI patched table. Put the CPIO archive in /boot and
+  # then enable the following lines
+  # boot.kernelParams = [
+  #   "mem_sleep_default=deep"
+  # ];
+  # boot.initrd.prepend = [
+  #   "/boot/acpi_override"
```

Comment:
this should probably be ``"${./path/to/acpi_override}"`` since nix will prepend this to initird

@azazel75 did you manage to get suspend working? 

once i apply this patch and uncomment above 2 lines i still don't see ``S3`` supported in acpi.
```
dmesg| grep -i "acpi: (supports"
[    0.188263] ACPI: (supports S0 S4 S5)
```

I also updated the script to generate patch to use `sed` since applying patch is broken with BIOS >= 1.24

Reply:
Hi @garbas, 
I can patch the line you commented, it's the same thing... in your case the absolute path is calculated by nix, I suppose!? (I'm a newbie)

About the patch, I had also to manually patch the "disassembled" code because the `patch` binary falied to apply some (final) hunks... I don't know which version my BIOS is at but I spawned up  Windows just three weeks ago and let it apply an upgrade to the BIOS using the installed Lenovo software. I was worried that it may break the ACPI S3 stuff but it still works after that.  Are you sure that the kernel patches the table on startup? My early kernel logs tell me that:
```
$ dmesg  | grep DSDT
[    0.000000] ACPI: DSDT ACPI table found in initrd [kernel/firmware/acpi/dsdt.aml][0x2338b]
[    0.000000] ACPI: Table Upgrade: override [DSDT-LENOVO-SKL     ]
[    0.000000] ACPI: DSDT 0x000000004FFC0000 Physical table override, new table: 0x000000004DF1E000
[    0.000000] ACPI: DSDT 0x000000004DF1E000 02338B (v02 LENOVO SKL      00000001 INTL 20180313)
[...]
$ dmesg | grep -i "acpi: (supports"
[    0.132924] ACPI: (supports S0 S3 S4 S5)
```

Reply:
Hi @garbas, I just checked my bios updates booting in Windows. It has upgraded my BIOS to 1.25 and S3 is still working

Reply:
@azazel75 i forgot to send you the link to the changes of the script i did -> https://github.com/fiji-flo/x1carbon2018s3/pull/1

Looks like suspend works for you.

regarding the `"/boot/acpi_override"`, during the build you can not access `/boot` but with `"${/boot/acpi_override}"` nix will prepend it to the file to `/nix/store` and then it can append it to initrd (https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/kernel/make-initrd.sh#L40).

being said that i'm still missing something since i dont see any override for the DSDT. need to fi

Reply:
I must say that I've hadn't any problem evaluating my configuration with "/boot/acpi_override", why do you say that you can not access `/boot`?
Yes I saw the line in  `make-initrd.sh` and with `/boot/acpi_override` as $PREP it should work. In fact, mine works flawlessly

Reply:
@azazel75 hmm, that is weird. it complained not to have permissions to open that file for me.

Here is my config: https://github.com/garbas/dotfiles/blob/master/nixos/grayworm.nix.
I'm running on nixos-unstable channel.